### PR TITLE
Feature/window frame pk

### PR DESCRIPTION
**Pull request by mustafasrepo**
## Which issue does this PR close?

Closes #.

### CODE REVIEW

File: datafusion/expr/src/window_frame.rs [hunk 1]
```diff
@@ -114,6 +114,21 @@ impl WindowFrame {
         }
     }
 
+    /// Creates a new window frame for OVER(ORDER BY <expr, ..>) cases.
+    pub fn new_with_primary_key_ordering() -> Self {
```
Comment (metesynnada) [inline datafusion/expr/src/window_frame.rs:118]: This constructor can be merged with new().

File: datafusion/physical-expr/src/window/window_expr.rs [hunk 2]
```diff
@@ -231,12 +231,13 @@ pub trait AggregateWindowExpr: WindowExpr {
...
         let mut row_wise_results: Vec<ScalarValue> = vec![];
+        let is_end_exact = self.get_window_frame().is_frame_end_exact();
         while idx < length {
             // Start search from the last_range. This squeezes searched range.
             let cur_range =
                 window_frame_ctx.calculate_range(&order_bys, last_range, length, idx)?;
-            // Exit if the range extends all the way:
-            if cur_range.end == length && not_end {
+            // Exit if the range extends all the way and end point is not exact:
+            if (cur_range.end == length && !is_end_exact) && not_end {
```
Comment (metesynnada) [inline datafusion/physical-expr/src/window/window_expr.rs:240]: Remove `( )`

File: datafusion/sql/src/expr/function.rs [hunk 3]
```diff
@@ -115,6 +131,8 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
             let window_frame = if let Some(window_frame) = window_frame {
                 regularize_window_order_by(&window_frame, &mut order_by)?;
                 window_frame
+            } else if is_ordering_stable {
```
Comment (metesynnada) [inline datafusion/sql/src/expr/function.rs:134]: This can be handled in the constructor.

**Review by metesynnada — APPROVED** [review]
LGTM.

**Comment by mustafasrepo** [comment]
Upstream [PR8842](https://github.com/apache/arrow-datafusion/pull/8842)

**Comment by mustafasrepo** [comment]
Merged upstream

**Pull request closed without merging**